File: packages/types/src/episodes.ts
/**
 * Canonical episode types — the dated things that happened to a repository.
 *
 * Canonical source: engine `EpisodeSummary` / `EpisodeDetail` /
 * `EpisodeListResponse` / `EpisodeCountsResponse` in
 * `packages/server/src/repowise/server/schemas/episodes.py`. Hand-written and
 * hand-kept in sync; nothing generates these.
 *
 * Two shapes because they cost different amounts. A summary is one indexed
 * SQLite read and carries no body; a detail spends one `git rev-list` to
 * answer whether the claim still holds. A list of summaries must never be
 * assembled by fetching details.
 */

/**
 * Only the tiers that describe the repository itself. The engine has a third,
 * `transcript`, which is per-machine and never crosses HTTP — two people
 * opening one dashboard would otherwise see different pages.
 */
export type EpisodeTier = "structural" | "git";

export interface EpisodeSummary {
  id: string;
  tier: EpisodeTier;
  /** Discriminates episodes within a tier, e.g. `code_fix`, `nested_repos`. */
  kind: string;
  subject: string;
  evidence: string;
  /** The bound scope, trimmed. `node_count` is the untrimmed total. */
  nodes: string[];
  /** Files the episode is bound to, before `nodes` was trimmed for display. */
  node_count: number;
  birth_commit: string | null;
  /** ISO-8601 UTC, or null when the store holds no date. */
  birth_at: string | null;
  last_seen_at: string | null;
  /**
   * The currency signal that costs no git call, or `null` on tiers that
   * accumulate members, where the same stamp proves nothing.
   *
   * `null` means *unchecked*, never *stale* — checking happens on the detail
   * route. Required-but-nullable rather than optional, because the engine
   * field carries a default and pydantic serializes defaults: the key is
   * always on the wire. Discriminate on the value, never on key presence.
   */
  still_true: string | null;
}

export interface EpisodeDetail {
  id: string;
  tier: EpisodeTier;
  kind: string;
  subject: string;
  body: string;
  evidence: string;
  nodes: string[];
  node_count: number;
  birth_commit: string | null;
  birth_at: string | null;
  last_seen_at: string | null;
  /** The sentence half of the verdict. Always present on a detail. */
  still_true: string;
  /**
   * The gate half. A reader asking *what happened here* shows the sentence
   * either way; anything putting a claim beside a statement about the present
   * must respect this.
   */
  current: boolean;
}

/**
 * `available: false` means this repository has never derived episodes — a
 * real cold start, not an error and not an empty store. Render it as
 * "nothing here yet", distinct from a store whose rows were all pruned.
 */
export interface EpisodeListResponse {
  available: boolean;
  /** Measured total behind the page, not the length of `episodes`. */
  total: number;
  episodes: EpisodeSummary[];
}

export interface EpisodeCountsResponse {
  available: boolean;
  total: number;
  /** Keys are a subset of the served tiers, so a consumer can switch on them. */
  by_tier: Partial<Record<EpisodeTier, number>>;
  by_kind: Record<string, number>;
}

File: packages/server/src/repowise/server/schemas/episodes.py
"""Episode response models — the dated things that happened to a repository.

Three shapes, and the split between them is a cost decision rather than a
presentational one. A **summary** carries no body and asks git nothing, so a
page of them is one indexed SQLite read. A **detail** carries the body and
spends the one ``git rev-list`` the currency answer needs. **Counts** are a
grouped aggregate, so a caller can state a total it measured instead of the
length of whatever page it happened to fetch.

Every response carries ``available``. A repository that has never derived
episodes is a normal 200 with ``available: false`` — not a 404, which would be
untrue (the repository exists), and not an empty list, which would be
indistinguishable from a repository whose episodes were all pruned. This
follows the sidecar-store convention the distill-savings endpoint already set.
"""

from __future__ import annotations

from datetime import UTC, datetime

from pydantic import BaseModel

#: Ceiling on the scope listed in a summary row. The measured worst case on
#: this repository's shareable tiers is 23 paths, and a sweep commit is allowed
#: to name a hundred, so the list is trimmed and ``node_count`` carries the
#: real total beside it. A cap without the count beside it is a silent lie
#: about coverage.
MAX_SUMMARY_NODES = 12


def _iso(ts: float | None) -> str | None:
    """An epoch stamp as ISO-8601 UTC, or None when there is no usable date.

    Every failure returns None rather than raising, and that is the whole
    point of the guard. ``birth_at`` is a bare ``REAL`` in an untyped SQLite
    column, so the response layer cannot assume any writer put a sane value
    there — an older repowise, an imported history with a year-10000 author
    date, or a hand-edited store are all enough. ``datetime.fromtimestamp``
    raises ``OSError`` outside ``time_t`` on Windows, ``OverflowError`` on
    infinities and ``ValueError`` on NaN, and an unguarded call takes down the
    whole page rather than the one row: a single bad stamp anywhere in the
    page would 500 the list.
    """
    if not ts:
        return None
    try:
        return datetime.fromtimestamp(ts, tz=UTC).isoformat()
    except (OSError, OverflowError, ValueError):
        return None


class EpisodeSummary(BaseModel):
    """One timeline row. No body, and no git call was made to build it."""

    id: str
    tier: str
    kind: str
    subject: str
    evidence: str
    # The scope, trimmed to MAX_SUMMARY_NODES. `node_count` is the untrimmed
    # total, so a reader can tell "bound to 3 files" from "bound to 3 of 85".
    nodes: list[str]
    node_count: int
    birth_commit: str | None
    birth_at: str | None
    last_seen_at: str | None
    # The currency signal that costs nothing: a structural fact re-observed by
    # a later index vouches for itself. Absent on every tier that accumulates
    # members, where the same stamp proves nothing — see `free_currency`. A
    # row without it is not stale, it is unchecked, and the detail endpoint is
    # where checking happens.
    still_true: str | None = None

    @classmethod
    def from_row(cls, row: dict, *, still_true: str | None = None) -> EpisodeSummary:
        nodes = [n for n in (row.get("nodes") or []) if isinstance(n, str) and n]
        return cls(
            id=row["id"],
            tier=row["tier"],
            kind=row["kind"],
            subject=row["subject"],
            evidence=row.get("evidence") or "",
            nodes=nodes[:MAX_SUMMARY_NODES],
            node_count=len(nodes),
            birth_commit=row.get("birth_commit"),
            birth_at=_iso(row.get("birth_at")),
            last_seen_at=_iso(row.get("last_seen_at")),
            still_true=still_true,
        )


class EpisodeDetail(BaseModel):
    """One episode, whole, with the currency question actually asked.

    ``current`` is the gate half of the verdict and ``still_true`` the sentence
    half. Both are served because they answer different questions: a reader
    asking *what happened here* wants the sentence even when the scope has
    moved, while anything putting a claim beside a statement about the present
    must respect the boolean.
    """

    id: str
    tier: str
    kind: str
    subject: str
    body: str
    evidence: str
    nodes: list[str]
    node_count: int
    birth_commit: str | None
    birth_at: str | None
    last_seen_at: str | None
    still_true: str
    current: bool

    @classmethod
    def from_row(cls, row: dict, *, sentence: str, current: bool) -> EpisodeDetail:
        nodes = [n for n in (row.get("nodes") or []) if isinstance(n, str) and n]
        return cls(
            id=row["id"],
            tier=row["tier"],
            kind=row["kind"],
            subject=row["subject"],
            body=row.get("body") or "",
            evidence=row.get("evidence") or "",
            nodes=nodes,
            node_count=len(nodes),
            birth_commit=row.get("birth_commit"),
            birth_at=_iso(row.get("birth_at")),
            last_seen_at=_iso(row.get("last_seen_at")),
            still_true=sentence,
            current=current,
        )


class EpisodeListResponse(BaseModel):
    """A page of summaries, with the measured total behind it."""

    available: bool = True
    total: int = 0
    episodes: list[EpisodeSummary] = []


class EpisodeCountsResponse(BaseModel):
    """Counts by tier and by kind, from a grouped read of the same filters.

    Exists for the same reason the decision counts endpoint does: a page that
    counts its own rows reports the size of a window, not the size of a store.
    """

    available: bool = True
    total: int = 0
    by_tier: dict[str, int] = {}
    by_kind: dict[str, int] = {}
